### ISSUE TITLE
[feature] Add a "data" package type. 

### ISSUE THREAD

**Issue by jasonbeach** [conan-io/conan#15379]
### What is your suggestion?

Add a "data" package type.

This is a continuation of a conversion started on slack:

https://cpplang.slack.com/archives/C41CWV9HA/p1703102258129209

Frequently when running a C++ program it is necessary to acquire data for applications. It could be map data, graphics / texture data, config files, etc. 

My specific use case is that my application requires a map of an area (robot search / path planning). So the application can run many times using the same map. There can be many areas that the robot operates in, so there can be many independent maps which should be easily selectable at runtime (i.e. when I run the software or reconfigure it while running, I tell it which map to use). 

My current workflow to manage the maps is that I have a folder on a server where all of the map data resides, and then when needed, I rsync that folder to the device that will run the application and then feed the root directory of the data to the application so it knows where to look for it. This is currently all  managed via bash scripts. When the application is run, I tell it which map I want it to use. 

The main disadvantage is that there's not a good way to only rsync the maps I'm interested in. I have to rsync all of them. yes you could specify folders  in the rsync command but that's a little clumsy and doesn't really scale. I would like to use conan for this as I think it would work really well for it. I've done some initial testing and think it works well.  

In addition to data that is required at runtime, I can also see use cases for data that is required during build (data that gets embedded in a binary), and data that is required at test time although I do not have specific experience with these two use-cases. 

The example `conanfile.py` I've used that seems promising is:
```
import os
from conan import ConanFile
from conan.tools.files import  load, copy


class ConanDataPackage(ConanFile):
  name = "data_package"
  license = "my license"
  author = "Jason Beach <jason.m.beach@gmail.com>"

  package_type = "application" # not ideal but sets traits so this package can be found at runtime. 
  exports_sources = "data/*", "version"

  def set_version(self):
    # read version from file if not provided on command line
    self.version = self.version or load(self, "version").strip()

  def package(self):
    copy(self, "*.csv", self.source_folder, self.package_folder)
    copy(self, "*.meta", self.source_folder, self.package_folder)

  def package_info(self):
    self.runenv_info.append_path("MAP_DATA_DIR", os.path.join(self.package_folder, "data"))

``` 

This package can be easily consumed by downstream packages like any other conan package.  The path to the data is exposed via the MAP_DATA_DIR environment variable (in this case)
### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by chfuerst** [comment]
I am using conan in my development, for also creating user interface. We write a Qt-based application, that operates on a user specific configuration (and different users get different views).

Would appreciate, and can fully confirm that the software package requires data interfaces, like icon files, default configuration template files, which can have impact on e.g. user interface.

Still, would seperate this config files into a seperate repository (for example in artifactory), which is maintained independent of e.g. third party software, but definitely see the benefit of fetching non-sources / binaries, and creating the release bundle / install interface.

So upvote from my side.

**Comment by memsharded** [comment]
Hi all.

Sorry @jasonbeach for not following up earlier.
Thanks @chfuerst for the feedback.

I understand the use case and the request. Let me start that there are already Conan users using Conan packages with only data in it. Even if it is not as "native/builtin", it seems totally doable.

> In addition to data that is required at runtime, I can also see use cases for data that is required during build (data that gets embedded in a binary), and data that is required at test time although I do not have specific experience with these two use-cases.

As @jasonbeach pointed out, the main challenge of the ``data`` package type is not defining a new type. It is:
- defining the derived ``requirement traits`` defaults
- defining how data requirements propagate down the graph, when they conflict or not (two different packages requiring ``data/1.0`` and ``data/2.0`` do conflict or not?
- defining ways to opt-in/opt-out the different conflict behavior
- Differentiating in build-time data packages and run-time data packages
- When a data package influences and effects the ``package_id`` of the consumers and when not (and how, in which mode). This is also connected to the data being embedded in binaries or just referenced for runtime

In summary the challenge is that there are many different usage patterns for "data" packages, so a single "data" new package-type would only be the tip of the iceberg of a much larger feature.

**Comment by chfuerst** [comment]
Fully agreed to your points @memsharded . I can just report about my use-case and my requirement, that i want to install files in a package. But i understand your concerns about contradicting versions in the tree. 

Further, i forgot to mention, that i use conan in a CMake environment, hence, i used the term "install" in the meaning of cmake / cpack install. 

Is it eventually possible to start with a minimal supported setup, like just copy commands? Or is it design strategy to first fully specify the detailed questions like "how to deal with conflicting versions", and only start to implement these board-agreed strategies once everybody is happy?

Thanks for your quick feedback,

BR; Christoph

**Comment by memsharded** [comment]
> Is it eventually possible to start with a minimal supported setup, like just copy commands? Or is it design strategy to first fully specify the detailed questions like "how to deal with conflicting versions", and only start to implement these board-agreed strategies once everybody is happy?

Yeah, the typical problem is that it is very easy to be trapped or blocked in that initial setup if other possible scenarios and use cases are not at least considered. I agree that it is not necessary to implement all the use cases at once, and we do most typically start with just 1 use case, but it is important to at least have some understanding of the other use cases.
